### PR TITLE
let range bloomfilter add index for fixed number of keys, instead of block

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: blockindex/bloomfilterindexer.go [hunk 1]
```diff
@@ -62,63 +69,80 @@ func NewBloomfilterIndexer(kv db.KVStore, cfg config.Indexer) (BloomFilterIndexe
...
+	if err != nil {
+		return err
+	}
+
+	if height > 0 {
+		bfx.curRangeBloomfilter, err = bfx.rangeBloomFilter(height)
+		if err != nil {
+			return err
+		}
+		bfx.currRangeBfKey, _ = bfx.totalRange.Get(height)
```
Comment (CoderZhi) [inline blockindex/bloomfilterindexer.go:124]: throw error

File: api/logfilter/logfilter.go [hunk 2]
```diff
@@ -205,25 +206,25 @@ func (l *LogFilter) SelectBlocksFromRangeBloomFilter(bf bloom.BloomFilter, start
 		}
 		if len(l.pbFilter.Topics) > 0 {
 			flag := false
+		checkTopic:
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply api/logfilter/logfilter.go:209]: current code used 2 break to break out of 2 for loop, see below at
https://github.com/iotexproject/iotex-core/pull/2563/files#diff-8ae2dae9d468db610c754871c3c4adddddb88ed3ccaba980ea7113023667d666L218

i think using a break tag is more clear and easier to understand

File: api/logfilter/logfilter.go [hunk 3]
```diff
@@ -205,25 +206,25 @@ func (l *LogFilter) SelectBlocksFromRangeBloomFilter(bf bloom.BloomFilter, start
...
-				if flag {
-					break
-				}
 			}
 			if !flag {
 				continue
 			}
 		}
-		blkNums = append(blkNums, blockHeight)
+		if blockHeight >= start && blockHeight <= end {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply api/logfilter/logfilter.go:225]: yeah, no need for this

File: blockindex/bloom_range.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+func (br *bloomRange) Bytes() ([]byte, error) {
+	if br.end < br.start {
+		return nil, errors.New("end should be larger than start")
+	}
+
+	b := br.BloomFilter.Bytes()
+	b = append(b, byteutil.Uint64ToBytesBigEndian(br.start)...)
+	b = append(b, byteutil.Uint64ToBytesBigEndian(br.end)...)
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockindex/bloom_range.go:41]: so bf.bytes is a big buffer, like ~100kB

if we do b = start + end, b = append(b, bf.bytes), it for sure incurs a re-allocation of underlying buffer

the current way is to hope b.Cap > b.Len + 16, then can avoid a re-allocation

File: blockindex/bloomfilterindexer.go [hunk 5]
```diff
@@ -62,63 +69,80 @@ func NewBloomfilterIndexer(kv db.KVStore, cfg config.Indexer) (BloomFilterIndexe
...
-		if tipHeight%bfx.rangeSize == 0 {
-			bfx.curRangeBloomfilter, _ = bloom.NewBloomFilter(bfx.bfSize, bfx.bfNumHash)
-		} else {
-			bfx.curRangeBloomfilter, err = bfx.rangeBloomFilter(tipHeight)
-			if err != nil {
-				return errors.Wrapf(err, "failed to read curRangeBloomfilter from DB")
-			}
+		tipHeight := byteutil.BytesToUint64BigEndian(tipHeightData)
+		if err = bfx.initRangeBloomFilter(tipHeight); err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockindex/bloomfilterindexer.go:94]: fixed

File: blockindex/bloomfilterindexer.go [hunk 6]
```diff
@@ -62,63 +69,80 @@ func NewBloomfilterIndexer(kv db.KVStore, cfg config.Indexer) (BloomFilterIndexe
...
 		}
 	case db.ErrNotExist:
-		if err = bfx.flusher.KVStoreWithBuffer().Put(RangeBloomFilterNamespace, []byte(CurrentHeightKey), byteutil.Uint64ToBytes(0)); err != nil {
+		if err = bfx.kvStore.Put(RangeBloomFilterNamespace, []byte(CurrentHeightKey), make([]byte, 8)); err != nil {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockindex/bloomfilterindexer.go:97]: fixed

File: blockindex/bloomfilterindexer.go [hunk 7]
```diff
@@ -62,63 +69,80 @@ func NewBloomfilterIndexer(kv db.KVStore, cfg config.Indexer) (BloomFilterIndexe
...
 			return err
 		}
-		if err := bfx.flusher.Flush(); err != nil {
-			return errors.Wrapf(err, "failed to flush")
+		if err = bfx.initRangeBloomFilter(0); err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockindex/bloomfilterindexer.go:101]: fixed

File: blockindex/bloomfilterindexer.go (hunk 1 above)
Comment (dustinxie) [reply blockindex/bloomfilterindexer.go:124]: fixed

**Review by CoderZhi — APPROVED** [review]

File: blockindex/bloom_range.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+func bloomRangeFromBytes(data []byte) (*bloomRange, error) {
+	length := len(data)
+	if length <= 16 {
+		return nil, errors.New("not enough data")
+	}
+
+	end := byteutil.BytesToUint64BigEndian(data[length-8:])
+	start := byteutil.BytesToUint64BigEndian(data[length-16 : length-8])
+	bf, err := bloom.BloomFilterFromBytes(data[:length-16])
```
Comment (CoderZhi) [inline blockindex/bloom_range.go:53]: I am strongly against this format.

File: api/logfilter/logfilter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (CoderZhi) [reply api/logfilter/logfilter.go:209]: I still believe it is not a good practice

**Review by CoderZhi — COMMENTED** [review]
Approving this PR, but I don't think we should have the following two implementations just for tiny efficiency tunings.

**Pull request merged**